### PR TITLE
RPC query account information and access keys by a specific block id

### LINKED ISSUE #2015: RPC query account information and access keys by a specific block id
*This issue supersedes #2008.*

Our clients want avoid the following race condition that can happen on RPC:

Suppose we want to query both the account info and the access keys of the account at the same time. Unfortunately, these are available through separate RPC requests. They can issue request for account state (suppose they get state at block X), but then when they query access keys get them at block X+Y. They try to re-query the state, but now it is at the block X+Y+Z, etc. This DevX does not seem nice, it would be better if they could issue a query targeting a specific block hash.

### PR DESCRIPTION
**Pull request by frol**
*Resolves #2015*

### CODE REVIEW

File: chain/jsonrpc/src/lib.rs [hunk 1]
```diff
@@ -267,16 +269,63 @@ impl JsonRpcHandler {
...
+                "account" => QueryRequest::ViewAccount { account_id },
+                "access_key" => match path_parts.len() {
+                    2 => QueryRequest::ViewAccessKeyList { account_id },
+                    3 => QueryRequest::ViewAccessKey {
+                        account_id,
+                        public_key: PublicKey::try_from(path_parts[2])
+                            .map_err(|_| RpcError::server_error(Some("Invalid public key")))?,
+                    },
+                    _ => {
+                        unreachable!("path parts are at least 2 and at most 3 elements due to splitn(3) and the check right after it");
```
Comment (bowenwang1996) [inline chain/jsonrpc/src/lib.rs]: maybe mention that this restriction is for `access_key` query?

File: core/primitives/src/views.rs [hunk 2]
```diff
@@ -165,7 +165,16 @@ pub enum QueryResponseKind {
     Error(QueryError),
     AccessKey(AccessKeyView),
     AccessKeyList(AccessKeyList),
-    Validators(EpochValidatorInfo),
```
Comment (bowenwang1996) [inline core/primitives/src/views.rs]: Why is this deleted?

Comment (frol) [reply core/primitives/src/views.rs]: This variant is unused :man_shrugging:

**Comment by frol** [comment]
@nearmax @bowenwang1996 This PR is ready to be merged.

**Comment by frol** [comment]
This PR introduced a new extended API which is backward-compatible (the old API of `["account/test.near", ""]` is preserved):

```
$ http post http://127.0.0.1:3030/ jsonrpc=2.0 id=dontcare \
    method=query \
    'params:={"request_type": "view_access_key_list", "account_id": "test.near", "block_id": 300 }'
```

We are heading away from the positional-based APIs for easier maintainability. For example, it will make adding `is_final`-like extensions (https://github.com/nearprotocol/nearcore/pull/1890/) easier and more natural.

File: chain/client/src/view_client.rs [hunk 3]
```diff
@@ -109,17 +110,26 @@ impl ViewClientActor {
...
-        let shard_id = self.runtime_adapter.account_id_to_shard_id(&account_id);
+
+        let header = match msg.block_id {
+            Some(BlockId::Height(block_height)) => self.chain.get_header_by_height(block_height),
+            Some(BlockId::Hash(block_hash)) => self.chain.get_block_header(&block_hash),
+            None => self.chain.head_header(),
+        };
+        let header = header.map_err(|e| e.to_string())?.clone();
+
+        let account_id = match &msg.request {
```
Comment (nearmax) [inline chain/client/src/view_client.rs]: I think it is a good call to not extract `account_id` to `RpcQueryRequest`

**Comment by nearmax** [comment]
Let's have a documentation issue for it :)

**Comment by frol** [comment]
@evgenykuzyakov GitHub says this PR requires your review before we merge

**Comment by frol** [comment]
Oh, I forgot to make the base64-encoding for the `Vec<u8>` fields. See #2034.

Let me make this change before we merge.

**Comment by frol** [comment]
I have fixed the API to handle base64-encoded strings instead of crazy arrays of numbers. To do that, I introduced two new types: `StoreKey` (I plan to introduce it throughout the code base down to `core/store/` in #2034) and `FunctionArgs` (I will create a separate PR to finish the migration to this type in the rest code base) to replace ad-hoc `Vec<u8>`, and to be able to do automatic base64 encoding with serde. I am planning to introduce even more named wrapping types to replace `Vec<u8>`.

@nearmax @bowenwang1996 @evgenykuzyakov Please, review.

File: chain/chain/src/test_utils.rs [hunk 4]
```diff
@@ -665,31 +666,27 @@ impl RuntimeAdapter for KeyValueRuntime {
...
-                    block_height,
-                    block_hash: *block_hash,
-                })
-            }
-            "access_key" if path.len() == 2 => Ok(QueryResponse {
+        match request {
+            QueryRequest::ViewAccount { account_id, .. } => Ok(QueryResponse {
+                kind: QueryResponseKind::ViewAccount(
+                    Account {
+                        amount: self.state.read().unwrap().get(&state_root).map_or_else(
```
Comment (evgenykuzyakov) [inline chain/chain/src/test_utils.rs]: map_or?

File: chain/client/src/view_client.rs [hunk 5]
```diff
@@ -109,17 +110,26 @@ impl ViewClientActor {
...
+            QueryRequest::ViewAccount { account_id, .. } => account_id,
+            QueryRequest::ViewState { account_id, .. } => account_id,
+            QueryRequest::ViewAccessKey { account_id, .. } => account_id,
+            QueryRequest::ViewAccessKeyList { account_id, .. } => account_id,
+            QueryRequest::CallFunction { account_id, .. } => account_id,
+        };
+        let shard_id = self.runtime_adapter.account_id_to_shard_id(account_id);
```
Comment (evgenykuzyakov) [inline chain/client/src/view_client.rs]: Do you need to check account ID validity before converting it to shard here?

File: core/primitives/src/serialize.rs [hunk 6]
```diff
@@ -76,6 +78,34 @@ pub trait BaseDecode: for<'a> TryFrom<&'a [u8], Error = Box<dyn std::error::Erro
     }
 }
 
+macro_rules! pretty_bytes_serialization {
```
Comment (evgenykuzyakov) [inline core/primitives/src/serialize.rs]: Why is it called `pretty`? Should it be `base64_bytes_serialization`?

File: chain/client/src/view_client.rs (hunk 5 above)
Comment (frol) [reply chain/client/src/view_client.rs]: Well, I would like to improve `AccountId` type, so it enforces the contract of being a valid account ID. Let's make it in another PR; created an issue for that: https://github.com/nearprotocol/nearcore/issues/2074

File: chain/chain/src/test_utils.rs (hunk 4 above)
Comment (frol) [reply chain/chain/src/test_utils.rs]: Well, that might be undesirable. Anyway, I have not touched this code, it is only the indentation that is changed in this line. We will need to address the `unwrap`s in a security review round.

File: core/primitives/src/serialize.rs (hunk 6 above)
Comment (frol) [reply core/primitives/src/serialize.rs]: Well, this is a reasonable comment, though, my thinking around this is that base64 is just an implementation detail, and the intent is to be able to serialize bytes into a "pretty format".
